cwlVersion: v1.1
class: CommandLineTool
baseCommand: [ethseq]

label: Computes the ethnicity of a sample using EthSEQ.

doc: |-
  EthSEQ provides an automated pipeline, implemented as R package, to annotate
  the ethnicity of individuals from WES data inspecting differential SNPs
  genotype profiles while exploiting variants covered by the specific assay. 

requirements:
  SchemaDefRequirement:
    types:
      - $import: "../types/ethseq_output_map.yaml"
  ShellCommandRequirement: {}
  InlineJavascriptRequirement: {}
  LoadListingRequirement:
    loadListing: shallow_listing

hints:
  DockerRequirement:
    dockerPull: demichelislab/ethseq:latest

inputs:
  model_gds_file:
    doc: |-
      The GDS file containing the model of the SNPs used for ethnicity
      inference.
    type: File
    inputBinding:
      position: 1
  genotype_vcf_file:
    doc: |-
      The VCF file containing the genotypes of the SNPs that are used for
      ethnicity inference.
    type: File
    inputBinding:
      position: 2
  output_directory_name:
    doc: The name of the folder where the EthSEQ output will be put.
    type: string?
    default: "ethseq"
    inputBinding:
      position: 3
  threads:
    doc: The number of parallel threads that will be used for computation.
    type: int?
    default: 1
    inputBinding:
      prefix: --num_threads
  enable_plot:
    doc: |-
      If true a graphical representation of the location of the sample in the
      PCA space where the ethnicity is inferred is created.
    type: boolean
    inputBinding:
      prefix: --enable_plot
  log_to_file:
    doc: |-
      If true the output generated by the tool either will be redirected to a
      file. Otherwise the output will be printed on the output.
    type: boolean
    default: true
  redirect_stdout_to_stderr:
    doc: |-
      If true includes the stderr output along with the stdout in the log file.
    type: boolean
    default: true
  log_filename:
    doc: The name of the output file that will contain the output.
    type: string
    default: "ethseq.log"

outputs:
  output:
    doc: The output directory produced by EthSEQ
    type: Directory?
    outputBinding:
      glob: "$(inputs.output_directory_name)"
  output_map:
    doc: |-
      A data structure that will allow easy access to the various outputs
      produced by ethseq.
    type: out_map:ethseq_output_map?
    outputBinding:
      glob: "$(inputs.output_directory_name)"
      outputEval: |
        ${
          var filt_fun = function (filt) { return function (ff) { return ff.basename.endsWith(filt) }; },
              files    = self[0].listing,
              out_map = {
                table:      files.filter(filt_fun(".txt"))[0],
                pca_coords: files.filter(filt_fun(".PCAcoord"))[0]
              };
          if (inputs.enable_plot) {
            out_map.report_pdf = files.filter(filt_fun(".pdf"))[0];
          }
          return out_map;
        }
  log_file:
    doc: |-
      The log file, if enabled, that captures the output produced by the tool.
    type: File?
    outputBinding:
      glob: $(inputs.log_filename)

arguments:
  - valueFrom: "$(inputs.log_to_file ? '2> ' + inputs.log_filename + (inputs.redirect_stdout_to_stderr ? ' 1>&2' : '') : '')"
    shellQuote: false
    position: 99999

$namespaces:
  out_map: "../types/ethseq_output_map.yaml#"
