cwlVersion: v1.1
class: CommandLineTool
baseCommand: [pacbam]

label: Computes the pileup of BAM files at specified positions

doc: |-
  PaCBAM is a C command line tool for the complete characterization of genomic
  regions and single nucleotide positions from next-generation sequencing data.
  PaCBAM implements a fast and scalable multi-core computational engine,
  generates exhaustive output files for downstream analysis, introduces an
  innovative on-the-fly read duplicates filtering strategy and provides
  comprehensive visual reports. 

requirements:
  ShellCommandRequirement: {}
  InlineJavascriptRequirement: {}

hints:
  DockerRequirement:
    dockerPull: demichelislab/pacbam:latest

inputs:
  bam_file:
    doc: The BAM file that contains the reads of the sample to analyze.
    type: File
    secondaryFiles: [^.bai, .bai]
    inputBinding:
      prefix: "bam="
      separate: false
  regions_bed_file:
    doc: The BED file containing the regions captured by the capture kit.
    type: File
    inputBinding:
      prefix: "bed="
      separate: false
  snps_vcf_file:
    doc: The VCF file containing the positions of the SNPs.
    type: File
    inputBinding:
      prefix: "vcf="
      separate: false
  reference_genome_fasta_file:
    doc: The file containing the reference genome in FASTA format.
    type: File
    secondaryFiles: [^.fai, .fai]
    inputBinding:
      prefix: "fasta="
      separate: false
  threads:
    doc: The number of parallel threads that will be used for computation.
    type: int
    default: 1
    inputBinding:
      prefix: "threads="
      separate: false
  min_base_quality:
    doc: Threshold that is used to exclude low-quality position within reads.
    type: int
    default: 20
    inputBinding:
      prefix: "mbq="
      separate: false
  min_read_quality:
    doc: Threshold that is used to exclude low-quality reads from counts.
    type: int
    default: 20
    inputBinding:
      prefix: "mrq="
      separate: false
  region_perc:
    doc: |-
      The fraction of the region to be considered when calculating the section
      with highest coverage.
    type: float
    default: 0.5
    inputBinding:
      prefix: "regionperc="
      separate: false
  log_to_file:
    doc: |-
      If true, the output generated by the tool will be redirected to a file.
      Otherwise the output will be printed on the output.
    type: boolean
    default: true
  redirect_stdout_to_stderr:
    doc: |-
      If true, it includes the stderr output along with the stdout in the log
      file.
    type: boolean
    default: true
  log_filename:
    doc: The name of the output file that will contain the output.
    type: string
    default: "pacbam.log"

arguments:
  - valueFrom: "2"
    prefix: "mode="
    separate: false
  - valueFrom: $(runtime.outdir)
    prefix: "out="
    separate: false
  - valueFrom: "$(inputs.log_to_file ? '2> ' + inputs.log_filename + (inputs.redirect_stdout_to_stderr ? ' 1>&2' : '') : '')"
    shellQuote: false
    position: 99999

outputs:
  output:
    doc: The pileup of the positions specified in the input files.
    type: File?
    outputBinding:
      glob: "*.snps"
  log_file:
    doc: |-
      The log file, if enabled, that captures the output produced by the tool.
    type: File?
    outputBinding:
      glob: $(inputs.log_filename)
